Mathlib/LinearAlgebra/Matrix/SchurTriangulation.lean

```diff
@@ -0,0 +1,300 @@
+eigenspace `V` is non-trivial. Given that `E` is also an inner product space, let `bV` and `bW` be
+othonormal bases for `V` and `Vᗮ` respectively. Then, the collection of vectors in `bV` and `bW`
+forms an othornomal basis `bE` for `E`, as the direct sum of `V` and `Vᗮ` is an internal
+decomposition of `E`. The matrix representation of `f` with respect to `bE` satisfies
+$$
+\sideset{_\mathrm{bE}}{_\mathrm{bE}}{[f]} =
+\begin{bmatrix}
+  \sideset{_\mathrm{bV}}{_\mathrm{bV}}{[f]} &
+  \sideset{_\mathrm{bW}}{_\mathrm{bV}}{[f]} \\
+  \sideset{_\mathrm{bV}}{_\mathrm{bW}}{[f]} &
+  \sideset{_\mathrm{bW}}{_\mathrm{bW}}{[f]}
+\end{bmatrix} =
+\begin{bmatrix} \mu I & □ \\ 0 & \sideset{_\mathrm{bW}}{_\mathrm{bW}}{[f]} \end{bmatrix},
+$$
+which is upper triangular as long as $\sideset{_\mathrm{bW}}{_\mathrm{bW}}{[f]}$ is. Finally, one
+observes that the recursion from $\sideset{_\mathrm{bE}}{_\mathrm{bE}}{[f]}$ to
+$\sideset{_\mathrm{bW}}{_\mathrm{bW}}{[f]}$ is well-founded, as the dimension of `bW` is smaller
+than that of `bE` because `bV` is non-trivial.
+
+However, in order to leverage `DirectSum.IsInternal.collectedOrthonormalBasis`, the type
+`Σ b, cond b (Fin m) (Fin n)` has to be used instead of the more natural `Fin m ⊕ Fin n` while their
+equivalence is propositionally established by `Equiv.sumEquivSigmalCond`.
+
+-/
+
+variable [IsAlgClosed 𝕜]
+
+/-- **Don't use this definition directly.** This is the key algorithm behind
+`Matrix.schur_triangulation`. -/
+protected noncomputable def SchurTriangulationAux.of {E : Type*} [NormedAddCommGroup E]
+    [InnerProductSpace 𝕜 E] [FiniteDimensional 𝕜 E] (f : Module.End 𝕜 E) :
+    SchurTriangulationAux f :=
+  haveI : Decidable (Nontrivial E) := Classical.propDecidable _
+  if hE : Nontrivial E then
+    let μ : f.Eigenvalues := default
+    let V : Submodule 𝕜 E := f.eigenspace μ
+    let W : Submodule 𝕜 E := Vᗮ
+    let m := Module.finrank 𝕜 V
+    have hdim : m + Module.finrank 𝕜 W = Module.finrank 𝕜 E := V.finrank_add_finrank_orthogonal
+    let g : Module.End 𝕜 W := orthogonalProjection W ∘ₗ f.domRestrict W
+    let ⟨n, hn, bW, hg⟩ := SchurTriangulationAux.of g
+
+    have bV : OrthonormalBasis (Fin m) 𝕜 V := stdOrthonormalBasis 𝕜 V
+    have hV := V.orthogonalFamily_self
+    have int : DirectSum.IsInternal (cond · V W) :=
+      suffices ⨆ b, cond b V W = ⊤ from (hV.decomposition this).isInternal _
+      (sup_eq_iSup V W).symm.trans Submodule.sup_orthogonal_of_completeSpace
+    let B (b : Bool) : OrthonormalBasis (cond b (Fin m) (Fin n)) 𝕜 ↥(cond b V W) := b.rec bW bV
+    let bE : OrthonormalBasis (Σ b, cond b (Fin m) (Fin n)) 𝕜 E :=
+      int.collectedOrthonormalBasis hV B
+    let e := Equiv.finAddEquivSigmaCond
+    let basis := bE.reindex e.symm
+    {
+      basis
+      dim := m + n
+      hdim := hn ▸ hdim.symm
+      upperTriangular := fun i j (hji : j < i) => show toMatrixOrthonormal basis f i j = 0 from
```

Comment:
Since this is a proof, it might be more idiomatic to use `by` to enter term mode and then replace `if` with `by_cases`; but this is only a style point, so unless another reviewer feels strongly it's probably ok as is.

Reply:
In my opinion, I find term mode structured with `if`, `show`, `suffices`, and `calc` much easier to read. I didn't even once use infoview during my development for this whole PR.

Reply:
> I didn't even once use infoview during my development of this whole PR.

Are you aware that the infoview works in term mode (under the "Expected Type" heading) as well as in tactic mode?

How did the development of this PR proceed if not using the expected type? Were you at least using the language server?

Reply:
Please revise this proof to use tactic mode. This is a requirement for readability and maintainability in Mathlib.